daemon/commands.c

```diff
@@ -0,0 +1,539 @@
+{
+    (void)index;
+}
+
+static void cmd_unlock_exclusive(unsigned index)
+{
+    (void)index;
+
+    uv_rwlock_wrunlock(&exclusive_rwlock);
+}
+
+static void cmd_unlock_orthogonal(unsigned index)
+{
+    uv_rwlock_rdunlock(&exclusive_rwlock);
+    uv_mutex_unlock(&command_lock_array[index]);
+}
+
+static void cmd_unlock_idempotent(unsigned index)
+{
+    (void)index;
+
+    uv_rwlock_rdunlock(&exclusive_rwlock);
+}
+
+static void cmd_unlock_high_priority(unsigned index)
+{
+    (void)index;
+}
+
+static void pipe_write_cb(uv_write_t* req, int status)
+{
+    (void)status;
+    uv_pipe_t *client = req->data;
+
+    uv_close((uv_handle_t *)client, NULL);
+    freez(client);
+    --clients;
+    info("Command Clients = %u\n", clients);
+}
+
+static inline void add_string_to_command_reply(char *reply_string, unsigned *reply_string_size, char *str)
+{
+    unsigned len;
+
+    len = strlen(str);
+    strncpyz(reply_string + *reply_string_size, str, len);
+    *reply_string_size += len;
+}
+
+static void send_command_reply(uv_stream_t *client, cmd_status_t status, char *message)
+{
+    int ret;
+    char reply_string[MAX_COMMAND_LENGTH] = {'\0', };
+    unsigned reply_string_size = 0;
+    uv_buf_t write_buf;
+    uv_write_t write_req;
+
+    add_string_to_command_reply(reply_string, &reply_string_size, cmd_status_str[status]);
+    if (message) {
+        add_string_to_command_reply(reply_string, &reply_string_size, ": ");
+        add_string_to_command_reply(reply_string, &reply_string_size, message);
+    }
+
+    write_req.data = client;
+    write_buf.base = reply_string;
+    write_buf.len = strlen((char *)write_buf.base);
+    ret = uv_write(&write_req, (uv_stream_t *)client, &write_buf, 1, pipe_write_cb);
+    if (ret) {
+        error("uv_write(): %s", uv_strerror(ret));
+    }
+    info("COMMAND: Sending reply: \"%s\"", reply_string);
+}
+
+cmd_status_t execute_command(cmd_t idx, char *args, char **message)
+{
+    cmd_status_t status;
+    cmd_type_t type = command_info_array[idx].type;
+
+    cmd_lock_by_type[type](idx);
+    status = command_info_array[idx].func(args, message);
+    cmd_unlock_by_type[type](idx);
+
+    return status;
+}
+
+static void after_schedule_command(uv_work_t *req, int status)
+{
+    struct command_context *cmd_ctx = req->data;
+
+    (void)status;
+
+    send_command_reply(cmd_ctx->client, cmd_ctx->status, cmd_ctx->message);
+    if (cmd_ctx->message)
+        freez(cmd_ctx->message);
+}
+
+static void schedule_command(uv_work_t *req)
+{
+    struct command_context *cmd_ctx = req->data;
+
+    cmd_ctx->status = execute_command(cmd_ctx->idx, cmd_ctx->args, &cmd_ctx->message);
+}
+
+static void parse_commands(uv_stream_t *client)
+{
+    char *message = NULL, *pos;
+    cmd_t i;
+    cmd_status_t status;
+    struct command_context *cmd_ctx;
+
+    status = CMD_STATUS_FAILURE;
+
+    for (i = 0 ; i < CMD_TOTAL_COMMANDS ; ++i) {
+        pos = strstr(command_string, command_info_array[i].cmd_str);
```

Comment:
Why is this a substring?

Reply:
It works for now but we probably have to adjust this when we add more commands.